File: Exercism Codes/PHP/Tournament.php
<?php

declare(strict_types=1);

class Tournament
{
    public function tally(string $scores): string {
        $data = $scores != '' ? array_map(function ($s) {
            return mb_split(";", $s);
        }, mb_split("\n", $scores)) : array();
        return $this->stringData($this->processData($data));
    }

    public function processData(array $data): array {
        $players = array();
        foreach($data as $d) {
            for($i = 0; $i < 2; $i++) {
                $status = $i == 0 ? $d[2] : $this->reverse($d[2]);
                $playerOneIndex = array_search($d[$i], array_map(function($p) {
                    return $p->name;
                }, $players));
                if($playerOneIndex === false) {
                    $players[] = new Player($d[$i], $status);
                }
                else {
                    $players[$playerOneIndex]->update($status);
                }
            }
        }
        usort($players, function($a, $b) {
            if ($a->points === $b->points) {
                return strcmp($a->name, $b->name);
            }
            return $b->points - $a->points;
        });
        return $players;
    }

    public function stringData(array $players): string {
        $line = array("Team                           | MP |  W |  D |  L |  P");
        foreach($players as $p) {
            $line[] = "$p->name".str_repeat(' ', 31 - strlen($p->name))."|  $p->played |  $p->win |  $p->draw |  $p->loss |  $p->points";
        }
        return implode("\n", $line);
    }

    public function reverse(string $status): string {
        return $status == "draw" ? "draw" : ($status == "win" ? "loss" : "win");
    }
}

class Player
{
    public string $name;
    public int $played;
    public int $win;
    public int $draw;
    public int $loss;
    public int $points;

    public function __construct(string $name, string $value) {
        $this->name = $name;
        $this->win = 0;
        $this->loss = 0;
        $this->draw = 0;
        $this->update($value);
        $this->compute();
    }

    public function update(string $value) {
        if($value == "win") $this->win++;
        if($value == "loss") $this->loss++;
        if($value == "draw") $this->draw++;
        $this->compute();
    }

    public function compute() {
        $this->played = $this->win + $this->loss + $this->draw;
        $this->points = ($this->win * 3) + $this->draw;
    }
}

File: Exercism Codes/Java/Tournament.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

class Tournament {

    private List<Player> playerData = new ArrayList<>();

    String printTable() {
        StringBuilder lines = new StringBuilder("Team                           | MP |  W |  D |  L |  P\n");
        for(Player p : playerData) {
            lines.append(String.format("%s%s|  %d |  %d |  %d |  %d |  %d\n", p.name, " ".repeat(31 - p.name.length()), p.played, p.win, p.draw, p.loss, p.points));
        }
        return lines.toString();
    }

    void applyResults(String resultString) {
        List<String[]> data = Arrays.stream(resultString.split("\n")).map(s -> s.split(";")).toList();
        List<Player> players = new ArrayList<>();
        for(String[] d : data) {
            for(int i = 0; i < 2; i++) {
                String status = i == 0 ? d[2] : reverse(d[2]);
                int playerIdx = players.stream().map(p -> p.name).toList().indexOf(d[i]);
                if(playerIdx < 0) players.add(new Player(d[i], status));
                else players.get(playerIdx).update(status);
            }
        }
        playerData = players.stream().sorted((a, b) -> a.points == b.points ? a.name.compareTo(b.name) : b.points - a.points).toList();
    }

    public String reverse(String status) {
        return status.equals("draw") ? "draw" : (status.equals("win") ? "loss" : "win");
    }

    static class Player {
        public String name;
        public int played, win, loss, draw, points;

        Player(String name, String value) {
            this.name = name;
            win = 0;
            loss = 0;
            draw = 0;
            update(value);
        }

        void update(String value) {
            if(value.equals("win")) win++;
            if(value.equals("draw")) draw++;
            if(value.equals("loss")) loss++;
            played = win + loss + draw;
            points = win * 3 + draw;
        }
    }
}
